### ISSUE TITLE
Pytorch eval mode

### ISSUE THREAD

**Issue by ehsankf** [Trusted-AI/adversarial-robustness-toolbox#484]
**Describe the bug**

**To Reproduce**

**Expected behavior**

With the models with BatchNorm2d in Pytorch, when attacking the model with eval mode, the performance of model is dramatically decreasing.

**Screenshots**

**System information (please complete the following information):**
 - OS
 - Python version
 - ART version or commit number
 - TensorFlow / Keras / PyTorch / MXNet version

**Comment by beat-buesser** [comment]
Hi @ehsankf Thank you very much for using ART! I'm interested to learn the reason for you observation. With performance do you mean accuracy or the time required to evaluate the model?

Would you have a minimal code example that reproduces your observation?

**Comment by ehsankf** [comment]
Thanks for your reply.
I mean the accuracy. I tried with resnet18. I applied the following work around after model.eval() and the performance 
improved:
```
for m in wrapped_model_ensemble._model.modules():
       if isinstance(m, nn.BatchNorm2d):
          m.track_running_stats=False
```

**Comment by beat-buesser** [comment]
There seem to be many discussions on this topic on https://discuss.pytorch.org/. What is the batch size during evaluation of your model?

**Comment by ehsankf** [comment]
I used 64. The benign accuracy does not change but for the attacks it provides significant degradation.

**Comment by ehsankf** [comment]
import sys
import argparse
import copy
import pdb
import os
import csv

import torch
import torch.nn as nn
import torch.nn.functional as F
import torch.optim as optim
import torchvision

import random
import numpy as np

from art.attacks import FastGradientMethod
from art.classifiers import PyTorchClassifier
from art.utils import load_cifar10



import models

if not torch.cuda.is_available():
   device = torch.device("cpu")
else:
   device = torch.device("cuda")


# Step 1: Load the MNIST dataset

(x_train, y_train), (x_test, y_test), min_pixel_value, max_pixel_value = load_cifar10()

# Step 1a: Swap axes to PyTorch's NCHW format

x_train = np.swapaxes(x_train, 1, 3).astype(np.float32)    
x_test = np.swapaxes(x_test, 1, 3).astype(np.float32)

pdb.set_trace()

def validate_classifier(Net):

    wrapped_model = PyTorchClassifier(
        Net,
        loss=nn.CrossEntropyLoss(),
        optimizer=torch.optim.SGD(Net.parameters(), lr=0.01),
        input_shape=(3, 32, 32),
        nb_classes=10,
        clip_values=(min_val, max_val)
    )

    
    #wrapped_model._model.eval()
   


 #fix
    '''
    for m in wrapped_model_ensemble._model.modules():
       if isinstance(m, nn.BatchNorm2d):
          m.track_running_stats=False
    '''

    predictions = wrapped_model.predict(x_test)

    accuracy = np.sum(np.argmax(predictions, axis=1) == np.argmax(y_test, axis=1)) / len(y_test)

    print("Benign accuracy of natural model on test set: {}%".format(accuracy * 100))

    attack = FastGradientMethod(classifier=wrapped_model, eps=eps_val)

    x_test_adv = attack.generate(x=x_test)

    predictions = wrapped_model.predict(x_test_adv)

    accuracy = np.sum(np.argmax(predictions, axis=1) == np.argmax(y_test, axis=1)) / len(y_test)
  
    print("Adversarial accuracy of natural model on test set: {}%".format(accuracy * 100))
    


DEVICE = device
NUM_CLASSES = 10
EPS = 0.03
model_name = 'ResNet18'
Net = models.__dict__[model_name]()


Net_path = 'pure_resnet18.pth'

checkpoint = torch.load(Net_path, map_location=device)

Net.load_state_dict(checkpoint)

min_val = 0
max_val = 1

eps_val = EPS 

validate_classifier(Net)

**Comment by ehsankf** [comment]
[](url)

**Comment by ehsankf** [comment]
[https://drive.google.com/file/d/1Uf3duaNsmuGBXH5vcdAPQsHhHnOQd2JW/view?usp=sharing](url)

**Comment by ehsankf** [comment]
```
'''ResNet in PyTorch.

For Pre-activation ResNet, see 'preact_resnet.py'.

Reference:
[1] Kaiming He, Xiangyu Zhang, Shaoqing Ren, Jian Sun
    Deep Residual Learning for Image Recognition. arXiv:1512.03385
'''
import torch
import torch.nn as nn
import torch.nn.functional as F

from torch.autograd import Variable


class BasicBlock(nn.Module):
    expansion = 1

    def __init__(self, in_planes, planes, stride=1):
        super(BasicBlock, self).__init__()
        self.conv1 = nn.Conv2d(in_planes, planes, kernel_size=3, stride=stride, padding=1, bias=False)
        self.bn1 = nn.BatchNorm2d(planes)
        self.conv2 = nn.Conv2d(planes, planes, kernel_size=3, stride=1, padding=1, bias=False)
        self.bn2 = nn.BatchNorm2d(planes)

        self.shortcut = nn.Sequential()
        if stride != 1 or in_planes != self.expansion*planes:
            self.shortcut = nn.Sequential(
                nn.Conv2d(in_planes, self.expansion*planes, kernel_size=1, stride=stride, bias=False),
                nn.BatchNorm2d(self.expansion*planes)
            )

    def forward(self, x):
        out = F.relu(self.bn1(self.conv1(x)))
        out = self.bn2(self.conv2(out))
        out += self.shortcut(x)
        out = F.relu(out)
        return out


class Bottleneck(nn.Module):
    expansion = 4

    def __init__(self, in_planes, planes, stride=1):
        super(Bottleneck, self).__init__()
        self.conv1 = nn.Conv2d(in_planes, planes, kernel_size=1, bias=False)
        self.bn1 = nn.BatchNorm2d(planes)
        self.conv2 = nn.Conv2d(planes, planes, kernel_size=3, stride=stride, padding=1, bias=False)
        self.bn2 = nn.BatchNorm2d(planes)
        self.conv3 = nn.Conv2d(planes, self.expansion*planes, kernel_size=1, bias=False)
        self.bn3 = nn.BatchNorm2d(self.expansion*planes)

        self.shortcut = nn.Sequential()
        if stride != 1 or in_planes != self.expansion*planes:
            self.shortcut = nn.Sequential(
                nn.Conv2d(in_planes, self.expansion*planes, kernel_size=1, stride=stride, bias=False),
                nn.BatchNorm2d(self.expansion*planes)
            )

    def forward(self, x):
        out = F.relu(self.bn1(self.conv1(x)))
        out = F.relu(self.bn2(self.conv2(out)))
        out = self.bn3(self.conv3(out))
        out += self.shortcut(x)
        out = F.relu(out)
        return out


class ResNet(nn.Module):
    def __init__(self, block, num_blocks, num_classes=10):
        super(ResNet, self).__init__()
        self.in_planes = 64

        self.conv1 = nn.Conv2d(3, 64, kernel_size=3, stride=1, padding=1, bias=False)
        self.bn1 = nn.BatchNorm2d(64)
        self.layer1 = self._make_layer(block, 64, num_blocks[0], stride=1)
        self.layer2 = self._make_layer(block, 128, num_blocks[1], stride=2)
        self.layer3 = self._make_layer(block, 256, num_blocks[2], stride=2)
        self.layer4 = self._make_layer(block, 512, num_blocks[3], stride=2)
        self.linear = nn.Linear(512*block.expansion, num_classes)

    def _make_layer(self, block, planes, num_blocks, stride):
        strides = [stride] + [1]*(num_blocks-1)
        layers = []
        for stride in strides:
            layers.append(block(self.in_planes, planes, stride))
            self.in_planes = planes * block.expansion
        return nn.Sequential(*layers)

    def forward(self, x):
        out = F.relu(self.bn1(self.conv1(x)))
        out = self.layer1(out)
        out = self.layer2(out)
        out = self.layer3(out)
        out = self.layer4(out)
        out = F.avg_pool2d(out, 4)
        out = out.view(out.size(0), -1)
        out = self.linear(out)
        return out


def ResNet18():
    return ResNet(BasicBlock, [2,2,2,2])

def ResNet34():
    return ResNet(BasicBlock, [3,4,6,3])

def ResNet50():
    return ResNet(Bottleneck, [3,4,6,3])

def ResNet101():
    return ResNet(Bottleneck, [3,4,23,3])

def ResNet152():
    return ResNet(Bottleneck, [3,8,36,3])


```

**Comment by ValerioB88** [comment]
Hi. Your issue is almost surely due to this problem: https://discuss.pytorch.org/t/batchnorm2d-bug-unintended-behaviour/112822
